src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,807 @@
+        except HTTPError as exc:
+            if exc.status == 404:
+                raise NoSuchObject(key)
+            else:
+                raise
+
+        return self._extract_b2_metadata(response, key)
+
+    @retry
+    @copy_ancestor_docstring
+    def get_size(self, key):
+        log.debug('started with %s', key)
+
+        try:
+            response = self._do_download_request('HEAD', key)
+        except HTTPError as exc:
+            if exc.status == 404:
+                raise NoSuchObject(key)
+            else:
+                raise
+
+        try:
+            return int(response.headers['Content-Length'])
+        except KeyError:
+            raise RuntimeError('HEAD request did not return Content-Length')
+
+    @retry
+    @copy_ancestor_docstring
+    def open_read(self, key):
+        try:
+            response = self._do_download_request('GET', key)
+        except NoSuchObject:
+            raise
+
+        try:
+            metadata = self._extract_b2_metadata(response, key)
+        except (BadDigestError, CorruptedObjectError):
+            # If there's less than 64 kb of data, read and throw
+            # away. Otherwise re-establish connection.
+            if response.length is not None and response.length < 64 * 1024:
+                self._get_download_connection().discard()
+            else:
+                self._close_connections()
+            raise
+
+        return ObjectR(key, response, self, metadata)
+
+    @prepend_ancestor_docstring
+    def open_write(self, key, metadata=None, is_compressed=False):
+        '''
+        The returned object will buffer all data and only start the uploads
+        when its `close` method is called.
+        '''
+
+        log.debug('started with %s', key)
+
+        headers = CaseInsensitiveDict()
+        if metadata is None:
+            metadata = dict()
+
+        self._add_b2_metadata_to_headers(headers, metadata)
+
+        return ObjectW(key, self, headers)
+
+    @retry
+    @copy_ancestor_docstring
+    def delete(self, key, force=False, is_retry=False):
+        log.debug('started with %s', key)
+
+        if self.disable_versions:
+            file_id, file_name = self._get_file_id_and_name(key)
+            file_ids = [{
+                'fileName': file_name,
+                'fileId': file_id
+            }]
+        else:
+            file_ids = self._list_file_versions(key)
+
+        if not file_ids:
+            raise NoSuchObject(key)
+
+        for file_id in file_ids:
+            try:
+                self._delete_file_id(file_id['fileName'], file_id['fileId'])
+            except B2Error as exc:
+                if exc.code == 'file_not_present':
+                    # Server may have deleted the object even though we did not
+                    # receive the response.
+                    if force or is_retry:
+                        pass
+                    else:
+                        raise exc
```

Comment:
`del file_ids[:i]` while you are iterating over `file_ids` does not work as expected. So in your current implementation https://github.com/s3ql/s3ql/blob/e5af517e66c52cd1a838cc9aa9d875f0c3cd2743/src/s3ql/backends/b2/b2_backend.py#L479-L486 the `if force or is_retry` case kinda breaks the iteration (and the else case is missing the `del`).

I would suggest rewriting it like this:

```python
    @retry
    def _delete_file_ids(self, file_ids, force=False, is_retry=False):
        for (i, file_id) in enumerate(file_ids):
            try:
                self._delete_file_id(file_id['fileName'], file_id['fileId'], force or is_retry)
            except:
                del file_ids[:i]
                raise

        del file_ids[:]

    def _delete_file_id(self, file_name, file_id, force=False):
        request_dict = {
            'fileName': file_name,
            'fileId': file_id
        }
        try:
            self._do_api_call('b2_delete_file_version', request_dict)
        except B2Error as exc:
            if exc.code == 'file_not_present' and force:
                return
            raise exc
        return

```

Here the force handling is moved in the `_delete_file_id` method so that deleting a non existing file version does not raise an error at all. Also after a retry the force parameter of `_delete_file_id` is always `True` since – as you mentioned in the comments – the server response might have been lost.

Reply:
Yeah, I tried to fix it yesterday and it seemed to work (at least with the test I wrote) but this is of course much better, thanks for the correction.